### PR TITLE
docs: blog post on parserOptions.project = true

### LINKED ISSUE #6520: Docs: Add post on parserOptions.project = true
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

As part of writing #6515, I wrote up some docs on using `parserOptions.project = true` (#6084). That feature isn't actually part of v6 - it's just something I'd wanted to get in before or in v6.

Since it's already landed in v5, let's add a blog post explaining it! It'll be good visibility for the enhancement.

<details>
<summary>Here's what I originally drafted...</summary>

### Feature: Relative `tsconfig.json` Projects

In typescript-eslint v5, in order to enable [typed linting](https://typescript-eslint.io/linting/typed-linting), your ESLint configuration must include a `parserOptions.project` property indicating which TSConfig(s) to use:

```js title=".eslintrc.cjs"
module.exports = {
  // ...
  parserOptions: {
    project: ['./tsconfig.json'],
  },
  // ...
};
```

This works well for repositories that have a single TSConfig for all files, such as small repositories using a single `tsconfig.json` or larger monorepos that have a `tsconfig.eslint.json` specifically used for linting.
But what if your project uses different TSConfig options for different TypeScript files in nested subdirectories?
Specifying the right array of relative TSConfig paths to try for each file in order can be cumbersome or downright impossible.

We recently added the option to specify `parserOptions.project` as just `true` in ESLint configurations.
This indicates to typescript-eslint that each file should be linted with the closest `tsconfig.json` file on disk.

```js title=".eslintrc.cjs"
module.exports = {
  // ...
  parserOptions: {
    project: true,
  },
  // ...
};
```

We hope using `project = true` simplifies configurations for repositories with nontrivial ESLint configuration setups.
We intend to add additional options around automatic TSConfig lookups to allow additional file names or other convenient behaviors.
</details>

### Affected URL(s)

https://typescript-eslint.io/blog/relative-parser-options-projects perhaps? Something like that?

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6520
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds a blog post explaining:

* The current state of `parserOptions.project` options 
* Why we added `project: true`
* Benefits of that new option
* How it works at a very high level
* Next steps

References:

* https://github.com/typescript-eslint/typescript-eslint/issues/101
* https://github.com/typescript-eslint/typescript-eslint/pull/6084

### CODE REVIEW

**Review by bradzacher — DISMISSED** [review]

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+It's even occasionally even reduced time spent on typed linting by helping projects use a simpler set of TSConfigs. 🚀
+
+In the long term, we're hoping to further improve the configuration and performance for typed linting (see _[Project Services](#project-services)_ below).
+Simplifying your configuration now will make it easier to onboard to our new options when they're available.
+
+### How It Works
+
+When `@typescript-eslint/parser` is configured to generate type information, it attaches a backing TypeScript "Program" for to file it parses.
+Those Programs provide type checking APIs used by lint rules.
+Each TSConfig file on disk is generally used to create exactly one Program, so files that refer to the same TSConfig file will reuse the same Program.
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:99]:
```suggestion
Each TSConfig file on disk is generally used to create exactly one Program, and files included by the same TSConfig file will reuse the same Program.
```

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+Depending on how the ESLint config's `parserOptions.project` was specified, determining _which_ TSConfig file to use for each file can be different:
+
+- For a single string (e.g. `"tsconfig.json"`), then only one Program will be created, and all linted files will reuse it.
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:103]:
```suggestion
- For a single path (e.g. `"tsconfig.json"`), only one Program will be created, and all linted files will reuse it.
```

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+- For globs and/or arrays (e.g. `"./packages/*/tsconfig.json"`), then each linted file will reuse a Program based on the _first_ matched TSConfig file.
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:104]:
```suggestion
- For globs and/or arrays (e.g. `"./packages/*/tsconfig.json"`), each linted file will use the Program created by the _first_ matched TSConfig file.
```

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+- The TypeScript compiler options used in the user's editor might differ from the compiler options in the TSConfigs they specified on disk.
+- Files not included in created Programs can't be linted with type information, even though editors still typically surface type information when editing those files.
+  - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).
+
+We're working on an option to instead call the same TypeScript "Project Service" APIs that editors such as VS Code use to create Programs for us instead.
+Project Services will automatically detect the TSConfig for each file (like `project: true`), and will also allow type information to be computed for JavaScript files without the `allowJs` compiler option (unlike `project: true`).
+
+We're hopeful this option will eventually become the standard way to enable typed linting.
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:144]:
```suggestion
We hope this option will eventually become the standard way to enable typed linting.
```

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+However, because it's so new and untested, we're keeping it under the `EXPERIMENTAL_` prefix for at least all of the `6.X` versions.
+
+See [Packages > Parser > `EXPERIMENTAL_useProjectService`](/packages/parser#EXPERIMENTAL_useProjectService) for more information.
+
+## Supporting typescript-eslint
+
+If you enjoyed this blog post and/or or use typescript-eslint, please consider [supporting us on Open Collective](https://opencollective.com/typescript-eslint).
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:151]:
```suggestion
If you enjoyed this blog post and/or use typescript-eslint, please consider [supporting us on Open Collective](https://opencollective.com/typescript-eslint).
```

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Some nits

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+- Glob pattern, such as `project: './packages/**/tsconfig.json'`
+- Array of paths and/or glob patterns, such as `project: ['./packages/**/tsconfig.json', './separate-package/tsconfig.json']`
+
+Explicitly indicating which TSConfig files are used for typed linting can be useful.
+Developers like being given explicit control over their tooling.
+However, we've seen a few issues arise from this approach:
+
+- Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)
+- Needing to change a template ESLint config every time it's used for a different repository structure is a pain
+- Using different TSConfigs than what your editor users can result in different lint reports in the editor verses on the command-line
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/website/blog/2023-08-02-parser-options-project-true.md:56]: Yeah if you run `eslint --config eslint.some.other.config.js` then you can give it totally different config values than what's in `eslint.config.js`.

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+:::
+
+### Why Try `true`
+
+If your project uses typed linting and manually specifies `tsconfig.json` files, we'd highly recommend trying out `parserOptions.project: true`.
+We've seen it reduce lines of codes of ESLint configurations in many projects that have switched to it.
+It's even occasionally even reduced time spent on typed linting by helping projects use a simpler set of TSConfigs. 🚀
+
+In the long term, we're hoping to further improve the configuration and performance for typed linting (see _[Project Services](#project-services)_ below).
+Simplifying your configuration now will make it easier to onboard to our new options when they're available.
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/website/blog/2023-08-02-parser-options-project-true.md:93]: I believe the current way is correct: one "onboards to" something.

**Review by bradzacher — DISMISSED** [review]

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+- Path, such as `project: './tsconfig.json'`
+- Glob pattern, such as `project: './packages/**/tsconfig.json'`
+- Array of paths and/or glob patterns, such as `project: ['./packages/**/tsconfig.json', './separate-package/tsconfig.json']`
+
+Explicitly indicating which TSConfig files are used for typed linting can be useful.
+Developers like being given explicit control over their tooling.
+However, we've seen a few issues arise from this approach:
+
+- Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:54]:
```suggestion
- Particularly large repos can end up with so many TSConfig globs, they become confusing to developers or even cause [performance issues from overly permissive globs](/linting/troubleshooting/performance-troubleshooting#wide-includes-in-your-eslint-options)
```

File: packages/website/blog/2023-08-02-parser-options-project-true.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+### Project Services
+
+The downside of having users specify `parserOptions.project` at all is that `@typescript-eslint/parser` needs manual logic to create TypeScript Programs and associate them with linted files.
+Manual Program creation logic comes with a few issues:
+
+- Complex project setups can be difficult to get right.
+  - For example, [typescript-eslint does not yet support Project References](https://github.com/typescript-eslint/typescript-eslint/issues/2094).
+- The TypeScript compiler options used in the user's editor might differ from the compiler options in the TSConfigs they specified on disk.
+- Files not included in created Programs can't be linted with type information, even though editors still typically surface type information when editing those files.
+  - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).
```
Comment (Josh-Cena) [inline packages/website/blog/2023-08-02-parser-options-project-true.md:138]:
```suggestion
  - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).
```

**Review by Josh-Cena — CHANGES_REQUESTED** [review]

**Review by Josh-Cena — APPROVED** [review]
LGTM

File: packages/website/blog/2023-09-12-parser-options-project-true.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+- Needing to change a template ESLint config every time it's used for a different repository structure is a pain
+- Using a TSConfig that's different from what your editor uses can result in different lint reports between the editor and the command-line
+
+Although developers may sometimes need exact control over their `parserOptions.project`, most of the time we just want to use the _nearest `tsconfig.json` to each linted file_, which is the TSConfig used by the editor by default.
+
+In other words, many developers want our [issue #101: Feature request: support looking up tsconfig.json relative to linted file](https://github.com/typescript-eslint/typescript-eslint/issues/101).
+
+## Introducing `true`
+
+As of typescript-eslint@5.52.0, we now support providing `true` for `parserOptions.project`:
```
Comment (armano2) [inline packages/website/blog/2023-09-12-parser-options-project-true.md:64]: `typescript-eslint@5.52.0` is marked as email

File: packages/website/blog/2023-09-12-parser-options-project-true.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+This post will explain what life was like before, what's changed, and what's coming next. 🎉
+
+<!--truncate-->
+
+## The Problem With Projects
+
+The `@typescript-eslint/parser` package is what enables ESLint to parse TypeScript source files.
+It converts raw TypeScript code into an ["AST" format](./2022-12-05-asts-and-typescript-eslint.md).
+When [`parserOptions.project`](https://typescript-eslint.io/packages/parser#project) is specified, it additionally sets up TypeScript programs that can be used by [typed rules](https://typescript-eslint.io/developers/custom-rules#typed-rules).
```
Comment (armano2) [inline packages/website/blog/2023-09-12-parser-options-project-true.md:29]:
```suggestion
When [`parserOptions.project`](/packages/parser#project) is specified, it additionally sets up TypeScript programs that can be used by [typed rules](/developers/custom-rules#typed-rules).
```

**Pull request merged**